addons/web/static/src/js/core/popover.js

```diff
@@ -26,37 +26,77 @@ odoo.define('web.Popover', function () {
             this.state = useState({
                 displayed: false,
             });
+
+            this._onClickDocument = this._onClickDocument.bind(this);
+            this._onScrollDocument = this._onScrollDocument.bind(this);
+            this._onResizeWindow = this._onResizeWindow.bind(this);
+
+            this._onScrollDocument = _.throttle(this._onScrollDocument, 50);
+            this._onResizeWindow = _.debounce(this._onResizeWindow, 250);
+        }
+
+        mounted() {
+            /**
+             * Use capture for the following events to ensure no other part of
+             * the code can stop its propagation from reaching here.
+             */
+            document.addEventListener('click', this._onClickDocument, {
+                capture: true,
+            });
+            document.addEventListener('scroll', this._onScrollDocument, {
+                capture: true,
+            });
+            window.addEventListener('resize', this._onResizeWindow);
+
+            this._compute();
         }
 
         patched() {
-            super.patched(...arguments);
-            this.show();
+            this._compute();
         }
 
+        willUnmount() {
+            document.removeEventListener('click', this._onClickDocument, {
+                capture: true,
+            });
+            document.removeEventListener('scroll', this._onScrollDocument, {
+                capture: true,
+            });
+            window.removeEventListener('resize', this._onResizeWindow);
+        }
+
+        //----------------------------------------------------------------------
+        // Private
+        //----------------------------------------------------------------------
+
         /**
-         * Show the popover according to its props. This method will try to position the
+         * Computes the popover according to its props. This method will try to position the
          * popover as requested (according to the `position` props). If the requested position
          * does not fit the viewport, other positions will be tried in a clockwise order starting
          * a the requested position (e.g. starting from left: top, right, bottom). If no position
          * is found that fits the viewport, 'bottom' is used.
+         *
+         * @private
          */
-        show() {
+        _compute() {
             if (!this.state.displayed) {
                 return;
             }
+
             // get the target from the dom
             // we don't want to do this early, since the target might not
-            // be in the dom when the component is instanciated
-            if (!this.constructor._isInViewport(this.el)) {
+            // be in the dom when the component is instantiated
+            if (!this.env.services.isElementInViewport(this.el)) {
```

Comment:
Not sure we want this in the env. Do you think of other usecases? Because for now, we can easily mock a static method on Popover in the tests to achieve what you want.

cc @ged-odoo What do you think?

Reply:
I believe mocking/patching has no place in the world of OWL/env.

If the same method is used both in the window instantiating the tests and in the tests themselves, it is not possible to patch one without affecting the other.

Reply:
> I believe mocking/patching has no place in the world of OWL/env.
 
I agree with you on that. However, I perceive tests a little bit differently. I believe it's acceptable to do that kind of stuff in tests, occasionally.

> If the same method is used both in the window instantiating the tests and in the tests themselves, it is not possible to patch one without affecting the other.

Not sure to understand what you mean, but if you mock a static method on Popover, it's very unlikely to be used by the test engine.

I'm not saying you shouldn't do it. As I said, I don't really know. But I'm not comfortable with the idea of having such functions in the env. I think that what bothers me is that it is obvious that i will be overridden to hack the result, not because it's something you want to configure in your app, or in a subtree.

Anyway, I'd like to have @ged-odoo opinion on this.

Reply:
> I think that what bothers me is that it is obvious that i will be overridden to hack the result, not because it's something you want to configure in your app, or in a subtree.

This is true, but if this ends up being the decisive criteria, note that it is also true for almost all the other things we put in the env services so far (the only exception seems to be `httpRequest` where we actually want a different behavior on the public side).

But I think tests should not be considered differently than configuring app. In regard to the components themselves, they should not be aware whether they are tested or not, they are just instantiated in another env, literally.

Using the `env` approach also make very explicit what is expected to change depending on the situation. With mock or patch, you never really know what is actually hacked or not, and it can make debugging (tests) issues harder (forgot to unpatch in afterEach which is impacting following tests, etc. None of this is necessary with env).

Reply:
@seb-odoo @aab-odoo 
I think that we should use an API that is designed exactly to solve this very issue: https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API.
This is also likely to be way more efficient in all cases.

Also, this is not really a service, but an access to a browser API. As such, I think that this should be located in `env.browser`.

- `env.browser` is where we add everything that is simply an access to native browser API, such as localStorage, Math.random, fetch, ...
- `env.services` is the location of higher level of abstraction pieces that needs to be alive for the duration of the application.  So, mail service for example. Services need to use `env.browser` as well if needed, so it can easily be mocked.

Reply:
This is interesting, but I'm not sure if it fits the need?

The current code does not really need an observer, as it doesn't need to be notified of element potentially getting out of view at some point in the future.

It is only checking if the currently computed position fits on the viewport at the particular time of the compute, to decide if another position has to be computed (if outside) or if it's fine to keep it as a valid position.
There's also the problem that if it's completely outside, no callback will ever be run apparently, except we don't want to wait indefinitely for a callback, we just want to know it's outside.

So changing that seems out of scope of the current PR, and maybe not needed at all.

With that in mind, what is left to do here? Make new `browser` env key and move the current method there instead of in `services`?